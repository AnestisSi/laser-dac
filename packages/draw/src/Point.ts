import { relativeToPosition, relativeToColor } from './helpers';

export type Color = [number, number, number];

export class Point {
  x: number;
  y: number;
  r: number;
  g: number;
  b: number;

  constructor(x: number, y: number, color?: Color) {
    this.x = relativeToPosition(x);
    this.y = relativeToPosition(y);
    this.r = color ? relativeToColor(color[0]) : 0;
    this.g = color ? relativeToColor(color[1]) : 0;
    this.b = color ? relativeToColor(color[2]) : 0;
  }
}
