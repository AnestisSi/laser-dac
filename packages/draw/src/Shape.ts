import { Point } from './Point';

export interface ShapeOptions {
  x: number;
  y: number;
}

export class Shape {
  x: number;
  y: number;

  constructor(options: ShapeOptions) {
    this.x = options.x;
    this.y = options.y;
  }

  draw(): Point[] {
    return [];
  }
}
