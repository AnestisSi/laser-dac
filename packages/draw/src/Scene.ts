import { Point } from './Point';
import { Shape } from './Shape';

const DEFAULT_FRAME_RATE = 15;
const DEFAULT_RESOLUTION = 500;

interface SceneOptions {
  // This number sets the requested number of points from a perpendicular line drawn from one side of the projection to the other.
  // Decreasing this number will make drawing faster but less accurate, increasing will make it slower but more accurate.
  resolution?: number;
}

export class Scene {
  points: Point[] = [];
  resolution: number;
  interval?: NodeJS.Timer;

  constructor(options?: SceneOptions) {
    this.resolution = (options && options.resolution) || DEFAULT_RESOLUTION;
  }

  add(...shapes: Shape[]) {
    shapes.forEach(shape => {
      const points = shape.draw(this.resolution);
      this.points = this.points.concat(points);
    });
  }

  reset() {
    this.points = [];
  }

  start(renderFrame: () => void, frameRate: number = DEFAULT_FRAME_RATE) {
    this.interval = setInterval(() => {
      this.reset();
      renderFrame();
    }, frameRate);
  }

  stop() {
    this.pause();
    this.reset();
  }

  pause() {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }
  }
}
