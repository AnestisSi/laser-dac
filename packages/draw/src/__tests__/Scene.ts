import { Scene } from '../Scene';
import { Shape } from '../Shape';
import { Point } from '../Point';

class MyShape extends Shape {
  x: number;
  y: number;
  constructor(options: { x: number; y: number }) {
    super();
    this.x = options.x;
    this.y = options.y;
  }
  draw() {
    return [new Point(this.x, this.y), new Point(this.x + 0.1, this.y + 0.1)];
  }
}

test('draw two shapes on the scene', () => {
  const scene = new Scene();
  const shapeA = new MyShape({ x: 0.1, y: 0.1 });
  const shapeB = new MyShape({ x: 0.5, y: 0.5 });
  scene.add(shapeA, shapeB);

  expect(scene.points).toMatchSnapshot();
});
