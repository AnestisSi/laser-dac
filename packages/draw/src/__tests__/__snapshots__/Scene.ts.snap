// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw a shape and use a transformer 1`] = `
Array [
  Point {
    "b": 0,
    "g": 0,
    "r": 0,
    "x": 26224,
    "y": 26214,
  },
  Point {
    "b": 0,
    "g": 0,
    "r": 0,
    "x": 19671,
    "y": 19661,
  },
]
`;

exports[`draw a shape on the scene 1`] = `
Array [
  Point {
    "b": 0,
    "g": 0,
    "r": 0,
    "x": 26214,
    "y": 26214,
  },
  Point {
    "b": 0,
    "g": 0,
    "r": 0,
    "x": 19661,
    "y": 19661,
  },
]
`;
